src/core/client.ts

```diff
@@ -733,40 +733,42 @@ export class Client<P = Indexable> implements Observable<ClientEvent<P>> {
 
       if (!realtimeSyncDocKeys.length) {
         logger.debug(`[WL] c:"${this.getKey()}" exit watch loop`);
-        return;
+        return Promise.resolve();
       }
 
       const req = new WatchDocumentsRequest();
       req.setClient(converter.toClient(this.id!, this.presenceInfo));
       req.setDocumentKeysList(realtimeSyncDocKeys);
 
-      const onStreamDisconnect = () => {
-        this.remoteChangeEventStream = undefined;
-        this.watchLoopTimerID = setTimeout(doLoop, this.reconnectStreamDelay);
-        this.eventStreamObserver.next({
-          type: ClientEventType.StreamConnectionStatusChanged,
-          value: StreamConnectionStatus.Disconnected,
+      return new Promise((resolve) => {
+        const onStreamDisconnect = () => {
+          this.remoteChangeEventStream = undefined;
+          this.watchLoopTimerID = setTimeout(doLoop, this.reconnectStreamDelay);
+          this.eventStreamObserver.next({
+            type: ClientEventType.StreamConnectionStatusChanged,
+            value: StreamConnectionStatus.Disconnected,
+          });
+          resolve();
```

Comment:
Could we call `reject()` here when `onStreamDisconnect` is called by `stream.on('error')`? If it is possible, we can represent the result of the `runWatchLoop` more precisely whether it is an error.